### DISCUSSION TITLE
Example front-ends setup for fast typing (tooltip on request)

### DISCUSSION THREAD

**Discussion by yugaego** [company-mode/company-mode#1214] [Show and tell]
Hey! I'll be the first one to post in this category 👩🏻‍🚀 🧑‍🚀 

I met this configurational approach in the issue #1116.

The idea is quite simple:
- **Show preview** of the **first available completion** inline.
  - Use <kbd>RET</kbd> to accept *currently previewed* completion.
  - Use <kbd>TAB</kbd> to accept the *common part* of all available completions (distinguished by the face).
  - Or keep typing.
- **Pop-up tooltip** only **if specifically requested**, where optionally:
  - <kbd>C-h</kbd> to lookup *documentation*.
  - <kbd>M-{digit}</kbd> to *quick-access* candidate.
  - <kbd>C-n</kbd> (or <kbd>M-n</kbd> on some setups) to *navigate* to the next candidate.
  - <kbd>ESC</kbd> to *hide* tooltip.

So, here's the **basic** setup (mostly copied from #1116):

```elisp
(require 'company)

(with-eval-after-load 'company

  (global-company-mode)

  (setq
   company-minimum-prefix-length 1
   company-idle-delay 0
   company-tooltip-idle-delay 10
   company-require-match nil
   company-frontends '(company-pseudo-tooltip-unless-just-one-frontend-with-delay
                       company-preview-frontend
                       company-echo-metadata-frontend)
   company-backends '(company-capf)
   company-show-quick-access t)

  (global-set-key (kbd "C-c c")
                  (lambda ()
                    (interactive)
                    (let ((company-tooltip-idle-delay 0.0))
                      (company-complete)
                      (and company-candidates
                           (company-call-frontends 'post-command))))))
```
![preview-2-2](https://user-images.githubusercontent.com/296460/133884393-22d290c2-978b-47ba-8097-f8cbcf88ad3d.gif)

![tooltip-2-2](https://user-images.githubusercontent.com/296460/133884482-48f6e7d3-427c-4620-80a9-e732b6ffef82.gif)

**Comment by dgutov** [comment]
Nice!

What if you hit `TAB` twice? Does it bring up the popup as well? That would seem like an easier way to trigger it.

**Reply by yugaego** [reply]
> What if you hit TAB twice? Does it bring up the popup as well? 

`TAB` completes common part with my setup. No popup. And I don't see a binding for double-tabs?

> That would seem like an easier way to trigger it.

With my keyboard/layout `TAB` requires more hand movement than `C-c c`, but having options is great.

BTW, it looks quite impressive to cycle right in the preview with 
```
(define-key company-active-map (kbd "<tab>") #'company-complete-common-or-cycle)
```

**Reply by dgutov** [reply]
> TAB completes common part with my setup. No popup. And I don't see a binding for double-tabs?

Ah, sorry. I had mistaken the mention of pressing `C-n` for one of the possible conditions for the popup to appear. It would generally imply that the popup comes up on TAB as well. But now that I think of it, it's probably a good thing that it doesn't automatically come up on `C-n`.

If someone does want it to appear on second `TAB` (or first one, depending on the binding), I think that can be arranged, though.

> (define-key company-active-map (kbd "<tab>") #'company-complete-common-or-cycle)

I like the binding and cycling is good. But when you get to cycling, perhaps it's time to show the popup too?

**Reply by yugaego** [reply]
> pressing C-n for one of the possible conditions for the popup to appear.

Gotcha.
[The coolest part of Emacs is that everything* (well, almost) is possible.]

> perhaps it's time to show the popup too?

Yep, maybe. The cycling looked fun on testing for the other thread, but it might turn out not to be very useful in the end (for my use cases). So will see.

**Reply by yugaego** [reply]
Dmitry, thank you for the idea to perform more actions on the same binding ❤️ 

(Though for my setup usability those actions would rather be bound to `C-c c` than `TAB`.)

I'm sure it's going to be refined/changed with time, but now the approach looks this way:


```elisp
  (defun yet-company-complete-common-or-raise-tooltip ()
    "Either insert the common part of all the candidates or raise a tooltip."
    (interactive)
    ;; Workaround to lookup completions from inside a word.
    (when (looking-at "\\S-")
      (save-excursion (insert " ")))
    (when (company-manual-begin)
      (let ((tick (buffer-chars-modified-tick)))
        (call-interactively 'company-complete-common)
        (when (eq tick (buffer-chars-modified-tick))
          (let ((company-tooltip-idle-delay 0.0))
            (company-complete)
            (and company-candidates
                 (company-call-frontends 'post-command)))))))

  (global-set-key (kbd "C-c c") #'yet-company-complete-common-or-raise-tooltip))
```

**Reply by dgutov** [reply]
Nice :-)

But here's a patch which adds the behavior I was thinking of:

```diff
diff --git a/company.el b/company.el
index d055a46..cfa2cb0 100644
--- a/company.el
+++ b/company.el
@@ -3585,6 +3585,7 @@ Delay is determined by `company-tooltip-idle-delay'."
   (cl-case command
     (post-command
      (if (or company-tooltip-timer
+             (company-explicit-action-p)
              (overlayp company-pseudo-tooltip-overlay))
          (if (not (overlayp company-preview-overlay))
              (company-pseudo-tooltip-unless-just-one-frontend command)
```

With that, you can go on typing -- no tooltip is displayed, only the preview -- and then invoke `company-complete-common` or `company-complete-common-or-cycle`. And right after that, if completion is still in progress, the popup will be displayed. So you can complete the unique match, or the common part, without it, but the popup will show up if pressing `TAB` (or whichever binding you prefer for "complete common") is not enough.

You can even unbind `C-n` and `C-p` in `company-active-map`, having these bindings free to move around. Or make them conditional on the popup being displayed:

```el
(defun my/filter-with-ptwd (cmd)
  (when (overlayp company-pseudo-tooltip-overlay)
    cmd))

(define-key company-active-map (kbd "C-n")
  '(menu-item "" company-select-next-or-abort :filter my/filter-with-ptwd))

(define-key company-active-map (kbd "C-p")
  '(menu-item "" company-select-previous-or-abort :filter my/filter-with-ptwd))
```

**Reply by dgutov** [reply]
I don't know if this behavior is something every user of this frontend would want, though.

**Reply by yugaego** [reply]
> can complete the unique match, or the common part, without it, but the popup will show up if pressing TAB (or whichever binding you prefer for "complete common") is not enough.

That would be a very neat one-liner patch. But during my tests tooltip is shown on `TAB` when completing a common part. Am I missing something?
Here's an example snippet (and the latest source with the added `(company-explicit-action-p)`) that I used:
```elisp
(setq company-minimum-prefix-length 1
      company-idle-delay 0
      company-tooltip-idle-delay 10
      company-frontends
      '(company-pseudo-tooltip-unless-just-one-frontend-with-delay
        company-preview-frontend)
      company-backends '(company-capf))
```
Then `deac| <TAB>` completes common part `deactivate-` and opens a tooltip with more candidates.


> Or make them conditional on the popup being displayed:
> I don't know if this behavior is something every user of this frontend would want, though.

Well, I think it's kinda cool 😉 (at least be aware of such possibility).

**Reply by dgutov** [reply]
> But during my tests tooltip is shown on TAB when completing a common part. Am I missing something?

That's the intended effect: "the popup will show up if pressing TAB (or whichever binding you prefer for "complete common") is not enough".

By pressing TAB you have indicated that you want to perform a completion action, but when `company-complete-common` results in multiple matches still, the popup is shown right away to help you choose.

Again, not sure if this behavior is going to be to everyone's liking, but it makes sense to me. At least in theory.

**Reply by yugaego** [reply]
> That's the intended effect ... when company-complete-common results in multiple matches still, the popup is shown right away to help you choose.

OK, I see.

> if this behavior is going to be to everyone's liking

It seems adding a way to force tooltip showing up with "tooltip-frontend-with-delay"
would be a nice (maybe even proper) thing to have out of the box.

At the moment, I don't have an opinion formed what could be the *exact* behavior of
such "force show of delayed tooltip".
So maybe let's wait and see if there's more interest in having this behavior, with a feedback on a preferred one.

Or consider one of the mentioned so far concepts (each can be refined if chosen):

1. 
```diff
diff --git a/company.el b/company.el
index d1ab486..67edbfa 100644
--- a/company.el
+++ b/company.el
@@ -2627,6 +2627,14 @@ inserted."
       (when company-candidates
         (setq this-command 'company-complete-common)))))
 
+(defun company-complete-and-show-delayed-tooltip ()
+  "Call `company-complete' and show tooltip with candidates."
+  (interactive)
+  (let ((company-tooltip-idle-delay 0.0))
+    (company-complete)
+    (and company-candidates
+         (company-call-frontends 'post-command))))
+
 (define-obsolete-function-alias
   'company-complete-number
   'company-complete-tooltip-row
```

2. 
```diff
diff --git a/company.el b/company.el
index d1ab486..9d9580f 100644
--- a/company.el
+++ b/company.el
@@ -2585,6 +2585,18 @@ With ARG, move by that many elements."
               (current-prefix-arg arg))
           (call-interactively 'company-select-next))))))
 
+(defun company-complete-common-or-show-delayed-tooltip ()
+  "Insert the common part of all candidates, or show a tooltip."
+  (interactive)
+  (when (company-manual-begin)
+    (let ((tick (buffer-chars-modified-tick)))
+      (call-interactively 'company-complete-common)
+      (when (eq tick (buffer-chars-modified-tick))
+          (let ((company-tooltip-idle-delay 0.0))
+            (company-complete)
+            (and company-candidates
+                 (company-call-frontends 'post-command)))))))
+
 (defun company-indent-or-complete-common (arg)
   "Indent the current line or region, or complete the common part."
   (interactive "P")
```

3. 
```diff
diff --git a/company.el b/company.el
index d1ab486..0bcd553 100644
--- a/company.el
+++ b/company.el
@@ -621,6 +621,12 @@ happens.  The value of nil means no idle completion."
                  (const :tag "immediate (0)" 0)
                  (number :tag "seconds")))
 
+(defcustom company-tooltip-cancel-delay-on-explicit-action nil
+  "On explicit completion action by user show delayed tooltip immediately.
+See `company-pseudo-tooltip-unless-just-one-frontend-with-delay'."
+  :type 'boolean
+  :package-version '(company . "0.9.14"))
+
 (defcustom company-begin-commands '(self-insert-command
                                     org-self-insert-command
                                     orgtbl-self-insert-command
@@ -3590,6 +3596,8 @@ Delay is determined by `company-tooltip-idle-delay'."
   (cl-case command
     (post-command
      (if (or company-tooltip-timer
+             (and company-tooltip-cancel-delay-on-explicit-action
+                  (company-explicit-action-p))
              (overlayp company-pseudo-tooltip-overlay))
          (if (not (overlayp company-preview-overlay))
              (company-pseudo-tooltip-unless-just-one-frontend command)
```

**Reply by yugaego** [reply]
> You can even unbind C-n and C-p in company-active-map, having these bindings free to move around. Or make them conditional on the popup being displayed:

This approach has inspired me to write the following *concept*, that would really make sense for `company-pseudo-tooltip-unless-just-one-frontend-with-delay` only. Posting it just in case here. Looks like frontends (or configurational modules similar to `tng`) potentially could have benefited having more separation similar to the backends.

```diff
diff --git a/company.el b/company.el
index d1ab486..64d7ac1 100644
--- a/company.el
+++ b/company.el
@@ -781,6 +781,13 @@ This is a default value of `company-quick-access-hint-function'."
   :type '(choice (const :tag "off" nil)
                  (const :tag "on" t)))
 
+(defcustom company-keymap-apply-filters nil
+  "Apply key bindings conditionally.
+Currently enables one filter function `company-keymap--filter-by-tooltip',
+called by `company-active-map' and `company-search-map' keymaps."
+  :type 'boolean
+  :package-version '(company . "0.9.14"))
+
 (defcustom company-async-redisplay-delay 0.005
   "Delay before redisplay when fetching candidates asynchronously.
 
@@ -800,14 +807,27 @@ asynchronous call into synchronous.")
 (defvar company-mode-map (make-sparse-keymap)
   "Keymap used by `company-mode'.")
 
+(defun company-keymap--filter-by-tooltip (command)
+  "Return a COMMAND if a tooltip is shown; otherwise return nil."
+  (when (or (not company-keymap-apply-filters)
+            (and (boundp 'company-pseudo-tooltip-overlay)
+                 (overlayp company-pseudo-tooltip-overlay)))
+    command))
+
 (defvar company-active-map
   (let ((keymap (make-sparse-keymap)))
     (define-key keymap "\e\e\e" 'company-abort)
     (define-key keymap "\C-g" 'company-abort)
     (define-key keymap (kbd "M-n") 'company--select-next-and-warn)
     (define-key keymap (kbd "M-p") 'company--select-previous-and-warn)
-    (define-key keymap (kbd "C-n") 'company-select-next-or-abort)
-    (define-key keymap (kbd "C-p") 'company-select-previous-or-abort)
+    (define-key keymap (kbd "C-n") '(menu-item
+                                     ""
+                                     company-select-next-or-abort
+                                     :filter company-keymap--filter-by-tooltip))
+    (define-key keymap (kbd "C-p") '(menu-item
+                                     ""
+                                     company-select-previous-or-abort
+                                     :filter company-keymap--filter-by-tooltip))
     (define-key keymap (kbd "<down>") 'company-select-next-or-abort)
     (define-key keymap (kbd "<up>") 'company-select-previous-or-abort)
     (define-key keymap [remap scroll-up-command] 'company-next-page)
@@ -828,7 +848,7 @@ asynchronous call into synchronous.")
     (define-key keymap "\C-s" 'company-search-candidates)
     (define-key keymap "\C-\M-s" 'company-filter-candidates)
     (company-keymap--bind-quick-access keymap)
-     keymap)
+    keymap)
   "Keymap that is enabled during an active completion.")
 
 (defvar company--disabled-backends nil)
```

**Reply by dgutov** [reply]
All of the above are interesting proposals.

Not sure how we're going to pick which ones to add to Company. Maybe after more people join the discussion and try them out?

Or feel free to open a PR with the ones you're going to use personally.

**Reply by yugaego** [reply]
>  Maybe after more people join the discussion and try them out?

Yes, let's wait for now.
And *might be* with time [files reorganization](https://github.com/company-mode/company-mode/pull/1154#issuecomment-890435398) would be an acceptable approach to follow, that would allow putting everything specifically "delayed tooltip" related into a separate file.

**Reply by yugaego** [reply]
> Or feel free to open a PR with the ones you're going to use personally.

Yep, that's what I did: #1273.

I also constantly use conditional bindings/filters (thanks to the great idea you had suggested) and added a record to [Wiki](https://github.com/company-mode/company-mode/wiki/Tips-&-tricks#enable-bindings-conditionally-for-a-shown-tooltip) about them. Frankly, I totally forgot there's the [extended version](https://github.com/company-mode/company-mode/discussions/1214#discussioncomment-1426069) of this approach.

After reviewing it now, it seems adding filters upstream might make sense if:
1. More use cases would benefit from it,
  - resulting in a better, more abstract solution from the start.

2. Company would define `company-menu`
  - also providing a way to set a filter for any key binding.

And I don't see a need for either of the two at this stage, so `Tips & Tricks` seems to be a good enough outcome (notifying the users that conditional bindings are possible).
